### PR TITLE
feat(permit): load pre generated permit info

### PR DESCRIPTION
**Pull request by alfetopito**
# Summary

Part of https://github.com/cowprotocol/token-lists/issues/293

Depends on https://github.com/cowprotocol/token-lists/pull/299

Use the pre-generated permit info data.

1. When checking whether the token is permittable, check first on the pre-generated data
- Only when that's not available, check the token is permittable locally
2. In the token selector, use a combination of pre-generated data and locally fetched permit info
- With that info, show the flag in the dropdown selector
image

image

# To Test

1. Open the token selector
* There should be tokens with the gas free flag
2. Pick a permittable token you don't have allowance for
3. Check the network tab and filter for `PermitInfo`
* There should be a request for `PermitInfo.<chain id>.json` for the current network you are on
* Switching to different tokens, to another tab, switching to a different window should NOT trigger a new request of this file
4. Place order
* Should pop up to sign permit
* Order appData should contain permit pre hook
5. Load another token that's permittable and is not in the list, such as `0x30d20208d987713f46dfd34ef128bb16c404d10f` (SD) on mainnet
6. Pick it as a sell token
* Permit info for it should be loaded and stored on `permittableTokens:v1` localStorage
7. Open the token selector and search for the token that was just added
* Should have the gas free flag.
6. Swap it
* Should work like the one from the list

### CODE REVIEW

File: apps/cowswap-frontend/src/modules/permit/const.ts [hunk 1]
```diff
@@ -9,3 +9,7 @@ export const ORDER_TYPE_SUPPORTS_PERMIT: Record<TradeType, boolean> = {
 }
 
 export const PENDING_ORDER_PERMIT_CHECK_INTERVAL = ms`1min`
+
+// TODO: obviously not final
+export const PRE_GENERATED_PERMIT_URL =
+  'https://raw.githubusercontent.com/cowprotocol/token-lists/bd247cad5146da28ff2fb4704e4ae444642a5535/src/public/PermitInfo'
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/permit/const.ts:15]: Waiting on https://github.com/cowprotocol/token-lists/pull/299 to be merged

File: apps/cowswap-frontend/src/modules/permit/hooks/useIsTokenPermittable.ts [hunk 2]
```diff
@@ -67,13 +81,25 @@ export function useIsTokenPermittable(
...
+    preGeneratedIsLoading,
+    provider,
+    spender,
+    tokenName,
+  ])
 
   if (isNative) {
     return false
   }
   // TODO: add an updater for this
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/permit/hooks/useIsTokenPermittable.ts:101]: Is the TODO still relevant?

File: apps/cowswap-frontend/src/modules/permit/hooks/usePreGeneratedPermitInfo.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+  const url = `${PRE_GENERATED_PERMIT_URL}.${chainId}.json`
+
+  const { data, isLoading } = useSWR(
+    url,
+    (url: string): Promise<Record<string, PermitInfo>> =>
+      fetch(url).then((r) => {
+        return r.json()
+      }),
+    {
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/permit/hooks/usePreGeneratedPermitInfo.ts:27]: I think it worth it to move the config in a const

**Review by shoom3301 — APPROVED** [review]

File: libs/common-const/src/common.ts [hunk 4]
```diff
@@ -166,6 +166,13 @@ export const SWR_OPTIONS = {
   revalidateOnFocus: false,
 }
 
+export const SWR_NO_REFRESH_OPTIONS = {
+  // Cache indefinitely
+  revalidateOnFocus: false,
+  revalidateOnReconnect: false,
+  refreshInterval: 0,
```
Comment (anxolin) [inline libs/common-const/src/common.ts:173]: will this cache also if it fails? or this caches only on successful requests?

File: apps/cowswap-frontend/src/modules/permit/hooks/usePreGeneratedPermitInfoForToken.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ * Fetch pre-generated permit info for a single token
+ */
+export function usePreGeneratedPermitInfoForToken(token: Nullish<Currency>): {
+  permitInfo: IsTokenPermittableResult
+  isLoading: boolean
+} {
+  const { allPermitInfo, isLoading } = usePreGeneratedPermitInfo()
+
+  const permitInfo =
+    allPermitInfo && token && 'address' in token ? allPermitInfo[token.address.toLowerCase()] : undefined
```
Comment (anxolin) [inline apps/cowswap-frontend/src/modules/permit/hooks/usePreGeneratedPermitInfoForToken.ts:19]: ubernit, my mind didn't compute the predicate at first, maybe we can save ' token && 'address' in token' into a const

pls, feel free to ignore, is not imp at all

**Review by anxolin — APPROVED** [review]

**Review by anxolin — COMMENTED** [review]
Beautiful ❤️, nice to see how all fits together: new token modal, script, permit check, library etc

File: libs/common-const/src/common.ts (hunk 4 above)
Comment (alfetopito) [reply libs/common-const/src/common.ts:173]: I assume it would only cache on success, but I have not tested.
Will give it a try.

Comment (alfetopito) [reply libs/common-const/src/common.ts:173]: Tested, and confirmed. It does keep trying when there's an error.

**Pull request merged**